File: apps/nihongo-android/app/src/test/java/com/edu/nihongo/utils/SrsAlgorithmTest.kt
package com.edu.nihongo.utils

import com.edu.nihongo.domain.entity.SrsCard
import com.edu.nihongo.domain.entity.SyncStatus
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class SrsAlgorithmTest {
    private fun card(
        easeFactor: Float = 2.5f,
        interval: Int = 0,
        repetitions: Int = 0,
        mastered: Boolean = false,
    ) = SrsCard(
        id = 1L,
        vocabularyId = 10L,
        easeFactor = easeFactor,
        interval = interval,
        repetitions = repetitions,
        nextReviewAt = System.currentTimeMillis(),
        mastered = mastered,
        syncStatus = SyncStatus.SYNCED,
        updatedAt = System.currentTimeMillis(),
    )

    @Test
    fun qualityBelowThree_resetsRepetitionsAndInterval() {
        val next = SrsAlgorithm.calculateNextReview(
            card(repetitions = 4, interval = 15),
            quality = 2,
        )
        assertEquals(0, next.repetitions)
        assertEquals(1, next.interval)
        assertFalse(next.mastered)
        assertEquals(SyncStatus.PENDING, next.syncStatus)
    }

    @Test
    fun clampsNegativeQualityAsAgain() {
        val next = SrsAlgorithm.calculateNextReview(
            card(repetitions = 3, interval = 10),
            quality = -2,
        )
        assertEquals(0, next.repetitions)
        assertEquals(1, next.interval)
    }

    @Test
    fun firstSuccess_usesIntervalOne() {
        val next = SrsAlgorithm.calculateNextReview(card(), quality = 4)
        assertEquals(1, next.repetitions)
        assertEquals(1, next.interval)
    }

    @Test
    fun secondSuccess_usesIntervalSix() {
        val next = SrsAlgorithm.calculateNextReview(
            card(repetitions = 1, interval = 1),
            quality = 4,
        )
        assertEquals(2, next.repetitions)
        assertEquals(6, next.interval)
    }

    @Test
    fun laterSuccess_multipliesByEase() {
        val next = SrsAlgorithm.calculateNextReview(
            card(repetitions = 2, interval = 6, easeFactor = 2.5f),
            quality = 4,
        )
        assertEquals(3, next.repetitions)
        assertEquals(15, next.interval)
    }

    @Test
    fun easeFactor_neverBelowMinimum() {
        val next = SrsAlgorithm.calculateNextReview(
            card(easeFactor = 1.3f),
            quality = 0,
        )
        assertTrue(next.easeFactor >= 1.3f)
    }

    @Test
    fun mastered_whenRepsAndIntervalMeetThreshold() {
        val next = SrsAlgorithm.calculateNextReview(
            card(repetitions = 4, interval = 21, easeFactor = 2.5f),
            quality = 5,
        )
        assertEquals(5, next.repetitions)
        assertTrue(next.interval >= 21)
        assertTrue(next.mastered)
    }

    @Test
    fun nextReviewAt_isAboutIntervalDaysAhead() {
        val before = System.currentTimeMillis()
        val next = SrsAlgorithm.calculateNextReview(card(), quality = 4)
        val dayMs = 86_400_000L
        assertTrue(next.nextReviewAt >= before + dayMs - 5_000)
        assertTrue(next.nextReviewAt <= System.currentTimeMillis() + dayMs + 5_000)
    }
}

File: apps/nihongo_flutter/test/srs_algorithm_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:nihongo_app/domain/entity/srs_card.dart';
import 'package:nihongo_app/domain/entity/sync_status.dart';
import 'package:nihongo_app/utils/srs_algorithm.dart';

SrsCard _card({
  double easeFactor = 2.5,
  int interval = 0,
  int repetitions = 0,
  bool mastered = false,
}) {
  final now = DateTime.now();
  return SrsCard(
    id: 1,
    vocabularyId: 10,
    easeFactor: easeFactor,
    interval: interval,
    repetitions: repetitions,
    nextReviewAt: now,
    mastered: mastered,
    syncStatus: SyncStatus.synced,
    updatedAt: now,
  );
}

void main() {
  group('SrsAlgorithm.calculateNextReview', () {
    test('clamps quality below 0 as again', () {
      final next = SrsAlgorithm.calculateNextReview(_card(repetitions: 3, interval: 10), -2);
      expect(next.repetitions, 0);
      expect(next.interval, 1);
      expect(next.syncStatus, SyncStatus.pending);
    });

    test('quality < 3 resets repetitions and sets interval to 1', () {
      final next = SrsAlgorithm.calculateNextReview(
        _card(repetitions: 4, interval: 15, easeFactor: 2.5),
        2,
      );
      expect(next.repetitions, 0);
      expect(next.interval, 1);
      expect(next.mastered, isFalse);
      expect(next.syncStatus, SyncStatus.pending);
    });

    test('first successful review uses interval 1', () {
      final next = SrsAlgorithm.calculateNextReview(_card(), 4);
      expect(next.repetitions, 1);
      expect(next.interval, 1);
    });

    test('second successful review uses interval 6', () {
      final next = SrsAlgorithm.calculateNextReview(
        _card(repetitions: 1, interval: 1),
        4,
      );
      expect(next.repetitions, 2);
      expect(next.interval, 6);
    });

    test('later reviews multiply interval by ease factor', () {
      final next = SrsAlgorithm.calculateNextReview(
        _card(repetitions: 2, interval: 6, easeFactor: 2.5),
        4,
      );
      expect(next.repetitions, 3);
      expect(next.interval, 15); // round(6 * 2.5)
    });

    test('ease factor never drops below 1.3', () {
      final next = SrsAlgorithm.calculateNextReview(
        _card(easeFactor: 1.3),
        0,
      );
      expect(next.easeFactor, greaterThanOrEqualTo(1.3));
    });

    test('mastered when repetitions >= 5 and interval >= 21', () {
      final next = SrsAlgorithm.calculateNextReview(
        _card(repetitions: 4, interval: 21, easeFactor: 2.5),
        5,
      );
      expect(next.repetitions, 5);
      expect(next.interval, greaterThanOrEqualTo(21));
      expect(next.mastered, isTrue);
    });

    test('schedules nextReviewAt about interval days ahead', () {
      final before = DateTime.now();
      final next = SrsAlgorithm.calculateNextReview(_card(), 4);
      final after = DateTime.now();

      expect(
        next.nextReviewAt.isAfter(before.add(const Duration(hours: 20))),
        isTrue,
      );
      expect(
        next.nextReviewAt.isBefore(after.add(const Duration(days: 2))),
        isTrue,
      );
    });
  });
}
